src/txn/txn.c

```diff
@@ -592,8 +590,28 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
 	    __wt_config_gets_def(session, cfg, "commit_timestamp", 0, &cval));
 	if (cval.len != 0) {
 #ifdef HAVE_TIMESTAMPS
-		WT_ERR(__wt_txn_parse_timestamp(
-		    session, "commit", &txn->commit_timestamp, &cval));
+		WT_ERR(__wt_txn_parse_timestamp(session, "commit", &ts, &cval));
+		/*
+		 * Commit timestamp being set should be newer than the oldest
+		 * timestamp and should move forward in a transaction.
+		 */
+		__wt_readlock(session, &txn_global->rwlock);
+		if (txn_global->has_oldest_timestamp && __wt_timestamp_cmp(&ts,
+		    &txn_global->oldest_timestamp) < 0)
+			ret = EINVAL;
+		__wt_readunlock(session, &txn_global->rwlock);
```

Comment:
I think this would be better as:

```c
WT_WITH_TIMESTAMP_READLOCK(session, &txn_global->rwlock,
   ret = (txn_global->has_oldest_timestamp && __wt_timestamp_cmp(&ts,
        &txn_global->oldest_timestamp) < 0) ? EINVAL : 0)
```